spark_log_parser/loaders/__init__.py

```diff
@@ -61,58 +71,110 @@ def __iter__(self):
                 self._add_chunk_to_size(chunk)
                 yield chunk
 
-    def _add_chunk_to_size(self, chunk):
+    def _add_chunk_to_size(self, chunk) -> None:
         self.total_size += len(chunk)
-        if self._maximum_allowed_size is not None and self.total_size > self._maximum_allowed_size:
+        if self.total_size > self._maximum_allowed_size:
             raise ArchiveTooLargeError()
 
-    def read(self, size=-1) -> bytes:
-        content = self._trailing_data
-        self._trailing_data = b''
+    def _num_bytes_left_in_buffer(self) -> int:
+        return len(self._buffer) - self._buffer_position
 
-        if not size or size == -1:
-            for chunk in self._chunks:
-                self._add_chunk_to_size(chunk)
-                content += chunk
-        else:
-            # iterate chunks until we've read at least `size` amount of data
+    def read(self, size=-1) -> bytes | bytearray | memoryview:
+        if self._consumed:
+            return b''
+
+        elif not size or size == -1:
             for chunk in self._chunks:
                 self._add_chunk_to_size(chunk)
-                content += chunk
-                if len(content) > size:
-                    break
+                self._buffer += chunk
 
-            # Trim off any data that is past the provided `size` and hold on to it for subsequent `read` calls
-            if len(content) > size:
-                self._trailing_data = content[size:]
-                content = content[:size]
+            self._consumed = True
+            return self._buffer
 
-        return content
+        else:
+            # Check if the data left in buffer if enough to satisfy the full read size...
+            if self._num_bytes_left_in_buffer() < size:
+                # If not, chop off everything that we've read out of the buffer, and start re-filling it. Reading
+                #  from our memoryview is actually slower here, since we end up having to convert that to bytes
+                #  anyway for our concatenation down below
+                self._buffer = self._buffer[self._buffer_position:]
```

Comment:
Do we want to copy bytes here? It's interesting that in this case we don't leave anything at the front of the buffer, but in the case below we leave everything before `start_pos`.

Reply:
> Do we want to copy bytes here?

I found that an intentional copy was the simplest thing to do here, yeah. When a `memoryview` is held on `bytearray` (or `bytes`, but those are immutable anyway), we can't modify that reference -

```python
bs = bytearray(b'asdoiufqwoieuh')
with memoryview(bs) as v:
    bs += b'foo'
    
Traceback (most recent call last):
  File "/Users/gseyffert/Library/Application Support/JetBrains/Toolbox/apps/PyCharm-P/ch-0/223.8214.51/PyCharm.app/Contents/plugins/python/helpers/pydev/pydevconsole.py", line 364, in runcode
    coro = func()
           ^^^^^^
  File "<input>", line 2, in <module>
BufferError: Existing exports of data: object cannot be re-sized
```

However, if we re-assign `self._buffer` to a new slice (copy), then we are free to append in more data into that copy -

```python
memview = memoryview(bs)
bs += b'foo'
Traceback (most recent call last):
  File "/Users/gseyffert/Library/Application Support/JetBrains/Toolbox/apps/PyCharm-P/ch-0/223.8214.51/PyCharm.app/Contents/plugins/python/helpers/pydev/pydevconsole.py", line 364, in runcode
    coro = func()
           ^^^^^^
  File "<input>", line 1, in <module>
BufferError: Existing exports of data: object cannot be re-sized

bs = bs[1:]
bs += b'foo'
bs
bytearray(b'sdoiufqwoieuhfoo')
```

Our old buffer & memoryview will then get GC'd at some point.

Reply:
> It's interesting that in this case we don't leave anything at the front of the buffer, but in the case below we leave everything before start_pos.

Re-assigning `self._buffer` on each `read` to lop off the already-consumed bytes was the main reason that the prior implementation was sooooo sloooooow. That is certainly more straightforward to do, logic-wise, but incurred that huge overhead as a result. This new implementation saves any copies/lopping off of data until we _know_ we need to read more data in. When we need to read in more data, we re-size our buffer in this manner to avoid the buffer getting too large. 

We could definitely recycle our buffer, and move the trailing bytes back to the front, and then start copying in new data to overwrite the old data. However, that seemed far more complicated than it is likely worth, especially since `read()` only accounts for like 0.3 - 0.4% (or less) of total CPU time now, at least based on the profiling I did.